### PR TITLE
Check that spline obeys joint limits before solving toppra

### PR DESCRIPTION
**Pull request by mpetersen94**
Changes in this PRs:
- Adds checks that the position limits for the spline are not violated.  If they are, additional waypoints are added to ride along the joint limit.

### TODOs / Nice-To-Haves
- ❌ Might want to tune resolution of joint limit violation checking

Checklists:
- [✅] Update HISTORY.md with a single line describing this PR

### CODE REVIEW

File: toppra/dracula/__init__.py [hunk 1]
```diff
@@ -146,6 +147,35 @@ def lims_obeyed(self, traj, raise_2nd_order):
...
+        lims_obeyed = True
+        t_max = 1 if multiplier is None else multiplier * self.t_sum
+        t_sample = np.arange(0, t_max, 0.001)
+        q_sample = traj(t_sample)
+
+        t_index = 0
+        waypt_copy = [self.waypts[0]]
+        for ii in range(1, self.waypts.shape[0]):
+            t_waypt = -1
+            closest_dist = 1e6
```
Comment (duanyutong) [inline toppra/dracula/__init__.py]: probably define closest_dist as a global const

File: toppra/dracula/__init__.py [hunk 2]
```diff
@@ -146,6 +147,35 @@ def lims_obeyed(self, traj, raise_2nd_order):
                 return False
         return True
 
+    def joint_limits_obeyed(self, traj, multiplier):
+        lims_obeyed = True
+        t_max = 1 if multiplier is None else multiplier * self.t_sum
```
Comment (duanyutong) [inline toppra/dracula/__init__.py]: if multiplier can be None, make it optional with default None

File: toppra/dracula/__init__.py [hunk 3]
```diff
@@ -146,6 +147,35 @@ def lims_obeyed(self, traj, raise_2nd_order):
...
+        t_sample = np.arange(0, t_max, 0.001)
+        q_sample = traj(t_sample)
+
+        t_index = 0
+        waypt_copy = [self.waypts[0]]
```
Comment (duanyutong) [inline toppra/dracula/__init__.py:158]: `waypt_copy` will be of the same shape as `waypts`, right? if shape is known, init with zeros or `waypts.copy()` to preallocate memory and modify its slices

File: toppra/dracula/__init__.py [hunk 4]
```diff
@@ -146,6 +147,35 @@ def lims_obeyed(self, traj, raise_2nd_order):
...
+
+            i, j = np.where(~(
+                (self.qlim[:, 0] < q_sample[t_index:t_waypt + 1]) &
+                (q_sample[t_index:t_waypt + 1] < self.qlim[:, 1])))
+            if i.size or j.size:
+                # t_index + i are violating knot points
+                lims_obeyed = False
+                waypt_copy += [(self.waypts[ii-1] + self.waypts[ii])/2.]
+            t_index = t_waypt
+            waypt_copy += [self.waypts[ii]]
```
Comment (duanyutong) [inline toppra/dracula/__init__.py]: `.append()` is the pythonic way or writing it but sill inefficient compared to modifying ndarray slices

File: toppra/dracula/__init__.py (hunk 3 above)
Comment (mpetersen94) [reply toppra/dracula/__init__.py:158]: No it won't.  It is at least as big as waypts but the amount bigger depends on the number of segments with joint limit violations.

**Review by 5yler — APPROVED** [review]
LGTM platform & style

**Pull request merged**